nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,186 @@
+ *
+ *  Copyright (c) 2020 Samsung Research Russia
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the <ORGANIZATION> nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Alexey Merzlyakov
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/keepout_filter.hpp"
+
+namespace nav2_costmap_2d
+{
+
+KeepoutFilter::KeepoutFilter()
+: costmap_filter_info_sub_(nullptr), map_filter_sub_(nullptr), map_filter_(nullptr)
+{
+}
+
+void KeepoutFilter::initializeFilter(
+  const std::string costmap_filter_info_topic)
+{
+  // Setting new costmap filter info subscriber
+  RCLCPP_INFO(
+    node_->get_logger(),
+    "Setting a subscriber to %s costmap filter info",
+    costmap_filter_info_topic.c_str());
+  costmap_filter_info_sub_ = node_->create_subscription<nav2_msgs::msg::CostmapFilterInfo>(
+    costmap_filter_info_topic, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::costmapFilterInfoCallback, this, std::placeholders::_1));
+
+  curr_time_ = node_->now();
+  prev_time_ = curr_time_;
+}
+
+void KeepoutFilter::costmapFilterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  // Resetting previous subscriber each time when new costmap filter information arrives
+  if (map_filter_sub_) {
+    RCLCPP_WARN(node_->get_logger(), "New map filter info arrived. Replacing old one.");
+    map_filter_sub_.reset();
+  }
+  map_filter_.reset();
```

Comment:
Don't reset this until you have something new to work with.

Reply:
There are two topics: filter info and mask topic that are not synchronized. If I will not reset `mask_` (formerly `map_filter_`) at the time when new filter info appears, this might cause a situation when we have new filter info and old map mask at the same time. This will lead to undefined behavior. Additionally `mask_` is used as indicator that everything is OK to work with. If `mask_==nullptr`, this means that filter need to wait.
So, I'd like to consider this `reset()` to not remove.

Reply:
If a new info topic came in with the new mask topic, I'd think you'd want to proceed with the old mask for as long as needed until the new one is received and ready to go. As is, it will reset so then the costmap update cycles that occur between getting the info topic and the mask topic will not include the keep out layer at all. That could cause the robot to navigate into a keepout zone and be forever stuck because there was a lapse of a few cycles to allow the robot to enter a bad spot.

Reply:
I don't think that's undefined behavior. You're waiting for the relevant information to come in. In the mask callback, you can then reset the mask and reassign it to its new value (under the layers' mutex, which I don't see since this is multithreaded). New info != enough to completely change over the state to it until you get the mask too. Any cycles where these layers "just go down" could be a real issue for a keep out or speed restricting zone where it misses critical behavior

Reply:
OK, this is reasonable. Thank you for explanation. After it, I would like agree that this `reset()` is redundant and should be removed. Before new map mask will arrive we will keep moving using old one.

> under the layers' mutex, which I don't see since this is multithreaded

Regarding multi-threads: there is something new. When I've made a map server, there were no cases for multi-thread usage. I've added locking guards for recursive costmap mutex into all call-backs as well as inside all API methods: `initializeFilter()`, `process()` and `resetFilter()`. Not sure, that these are not an extra-measures. Could you please share use-cases of multi-thread using of costmap plugins?

Reply:
BTW, searching for `mutex` keyword in plugins directory giving me only one plugin is locking its resources: StaticLayer. Strange, that other plugins are not prepared for multi-thread cases.

Reply:
They all have locks. See https://github.com/ros-planning/navigation2/blob/master/nav2_costmap_2d/plugins/obstacle_layer.cpp#L483 the data buffer has locks.  Static layer would be a good analog to model though.

Reply:
Any callback could be done at the same time as the update thread so you need to lock the resources that are used in updates when in the callbacks

Reply:
I've checked the shared with callbacks resources for KeeputFilter: `mask_sub_`, `mask_topic_`, `filter_info_topic_` and `mask_costmap_`. They are all used by API methods, so my initial suppose about multi-thread locking was correct.
Added locks to API methods and two callbacks.